docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,793 @@
+  - "@sbueringer"
+reviewers:
+  - "@enxebre"
+  - "@vincepri"
+  - "@CecileRobertMichon"
+creation-date: 2022-02-21
+last-updated: 2022-02-21
+status: implementable
+see-also:
+  -
+replaces:
+  -
+superseded-by:
+  -
+---
+
+# Cluster API Runtime SDK
+
+## Table of Contents
+
+- [Cluster API Runtime SDK](#cluster-api-runtime-sdk)
+  - [Table of Contents](#table-of-contents)
+  - [Glossary](#glossary)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+    - [Future Work](#future-work)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      - [Cluster API Runtime Hooks vs Kubernetes admission webhooks](#cluster-api-runtime-hooks-vs-kubernetes-admission-webhooks)
+      - [Runtime SDK rules](#runtime-sdk-rules)
+  - [Runtime Extensions developer guide](#runtime-extensions-developer-guide)
+    - [Implementing Runtime Extensions](#implementing-runtime-extensions)
+      - [Timeouts](#timeouts)
+      - [Availability](#availability)
+      - [Side Effects](#side-effects)
+      - [Idempotence](#idempotence)
+      - [Avoid dependencies](#avoid-dependencies)
+    - [Deploy Runtime Extensions](#deploy-runtime-extensions)
+      - [Availability](#availability-1)
+      - [Identity and access management](#identity-and-access-management)
+      - [Alternative deployments methods](#alternative-deployments-methods)
+    - [Registering Runtime Extensions](#registering-runtime-extensions)
+  - [Runtime Hooks developer guide (CAPI internals)](#runtime-hooks-developer-guide-capi-internals)
+    - [Runtime hook implementation](#runtime-hook-implementation)
+    - [Discovering Runtime Extensions](#discovering-runtime-extensions)
+    - [Calling Runtime Extensions](#calling-runtime-extensions)
+  - [Security Model](#security-model)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+    - [Version Skew Strategy](#version-skew-strategy)
+  - [Annex](#annex)
+    - [Runtime SDK rules](#runtime-sdk-rules-1)
+  - [Implementation History](#implementation-history)
+  
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+- **Cluster API Runtime**: identifies the Cluster API execution model, a set of controllers cooperating in managing the 
+  Workload Cluster’s lifecycle.
+- **Runtime SDK**: a set of rules, recommendations and fundamental capabilities required to develop Runtime Hooks and 
+  Runtime Extensions.
+- **Runtime Hook**: a single, well identified, extension point allowing applications built on top of Cluster API to hook
+  into specific moments of the Workload Cluster’s lifecycle, e.g. Cluster.BeforeUpgrade.
```

Comment:
>into specific moments of the Workload Cluster’s lifecycle

Does this include both Cluster and Machine CRs or something different?

Reply:
This includes everything in a Cluster, including the control plane and worker machines, but eventually, also MHC and other resources are relevant for the Cluster lifecycle

Reply:
Added an example to clarify this